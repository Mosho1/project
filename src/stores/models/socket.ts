import { values } from 'mobx'
import { types, getParent } from 'mobx-state-tree'
import { randomUuid } from '../utils/utils';
import { modelTypes } from './index';

const socketType = types.enumeration('socketType', [
    'input',
    'output',
    'execInput',
    'execOutput',
]);

export type SocketTypeEnum = typeof socketType.Type;

export const socketTypes = (socketType as any).types.map((x: any) => x.value) as SocketTypeEnum[];

export const areSocketsCompatible = (s1: SocketType, s2: SocketType) => {
    if (s1.isInput === s2.isInput) return false;
    if (s1.isExec !== s2.isExec) return false;
    const [input, output] = s1.isInput ? [s1, s2] : [s2, s1];
    if (input.isExec && (input.arrows.length > 0 || output.arrows.length > 0)) return false;
    if (input.arrows.length > 0) return false;
    return true;
};

export const Socket = types.model('Socket', {
    id: types.optional<string, string>(types.identifier(), randomUuid) as any as string,
    socketType,
    name: types.optional(types.string, '')
})
    .views(self => ({
        get box(): modelTypes['Box'] {
            return getParent(self, 2);
        },
    }))
    .views(self => ({
        get index() {
            const { inputs, outputs, execInputs, execOutputs } = self.box;
            switch (self.socketType) {
                case 'input':
                    return execInputs.length + inputs.findIndex(s => s === self);
                case 'output':
                    return execOutputs.length + outputs.findIndex(s => s === self);
                case 'execInput':
                    return execInputs.findIndex(s => s === self);
                case 'execOutput':
                    return execOutputs.findIndex(s => s === self);
            }
            return 0;
        },
    }))
    .views(self => ({
        get isExec() {
            return self.socketType === 'execInput' || self.socketType === 'execOutput';
        },
        get isInput() {
            return self.socketType === 'execInput' || self.socketType === 'input';
        },
        get arrows() {
            if (!self.box.store) return [];
            return values(self.box.store.arrows).filter(a => a.input === self || a.output === self);
        },
        get x() {
            if (self.socketType === 'input' || self.socketType === 'execInput') {
                return self.box.x + 18;
            } else {
                return self.box.x + self.box.width - 18;
            }
        },
        get y() {
            return 50 + self.box.y + self.index * 30;
        }
    }))
    .actions(self => {
        return {
            setName(name: string | null) {
                self.name = name || '';
            },
            isCompatibleWith(socket: typeof self) {
                return areSocketsCompatible(self as any, socket as any);
            }
        }
    });

export type SocketType = typeof Socket.Type;
export type SocketSnapshotType = typeof Socket.SnapshotType;
