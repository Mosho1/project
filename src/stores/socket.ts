// eslint-disable-next-line
import {
    onSnapshot,
    applySnapshot,
} from "mobx-state-tree"

let subscription: any;
export default function syncStoreWithBackend(socket: any, store: any) {
    // === SYNC PATCHES (recommended)
    // subscription = onPatch(store, (data) => {
    //     socketSend(data)
    // })

    // onSocketMessage((data) => {
    //     applyPatch(store, data)
    // })

    // === SYNC ACTIONS
    // subscription = onAction(store, data => {
    //     socketSend(data)
    // })

    // onSocketMessage(data => {
    //     applyAction(store, data)
    // })

    // === SYNC SNAPSNOTS
    subscription = onSnapshot(store, (data) => {
        socketSend(data)
    })

    onSocketMessage((data: any) => {
        applySnapshot(store, data)
    })

    let isHandlingMessage = false
    function socketSend(data: any) {
        if (!isHandlingMessage) socket.send(JSON.stringify(data))
    }

    function onSocketMessage(handler: any) {
        socket.onmessage = (event: any) => {
            isHandlingMessage = true
            handler(JSON.parse(event.data))
            isHandlingMessage = false
        }
    }
}

/**
 * Clean up old subscription when switching communication system
 */
if (module.hot) {
    module.hot.dispose(() => {
        subscription();
    })
}
