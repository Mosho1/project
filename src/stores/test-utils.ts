import { mapValues } from 'lodash';

export const mock = <T>(instance: T, mock: NestedPartial<T>) => {
    for (let k in mock) {
        const descriptor = Object.getOwnPropertyDescriptor(instance, k);
        Object.defineProperties(instance, mapValues(mock, value => ({
            value,
            enumerable: descriptor!.enumerable,
            configurable: true,
            writable: true,
        })));
    }
    return instance;
};

const f = (a: any[], b: any[]): any[] =>
    [].concat(...a.map(a2 => b.map((b2: any) => [].concat(a2 as any, b2 as any) as any) as any));

export const product = (a: any[], b: any[], ...c: any[]): any[][] => {
    if (!b || b.length === 0) {
        return a;
    }
    const [b2, ...c2] = c;
    const fab = f(a, b);
    return product(fab, b2, ...c2);
};