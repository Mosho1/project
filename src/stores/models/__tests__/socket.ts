import { Socket, SocketTypeEnum, socketTypes, areSocketsCompatible } from '../socket';
import { mock, product } from '../../test-utils';
import { ObservableMap } from 'mobx';

test('isExec', () => {
    expect(Socket.create({ socketType: 'input' }).isExec).toBe(false);
    expect(Socket.create({ socketType: 'output' }).isExec).toBe(false);
    expect(Socket.create({ socketType: 'execInput' }).isExec).toBe(true);
    expect(Socket.create({ socketType: 'execOutput' }).isExec).toBe(true);
});

test('isInput', () => {
    expect(Socket.create({ socketType: 'input' }).isInput).toBe(true);
    expect(Socket.create({ socketType: 'output' }).isInput).toBe(false);
    expect(Socket.create({ socketType: 'execInput' }).isInput).toBe(true);
    expect(Socket.create({ socketType: 'execOutput' }).isInput).toBe(false);
});

describe('index', () => {
    let box: { inputs: any[], outputs: any[], execInputs: any[], execOutputs: any[] };
    beforeEach(() => {
        box = {
            inputs: [],
            outputs: [],
            execInputs: [],
            execOutputs: [],
        };
    });

    test('input', () => {
        const socket = Socket.create({ socketType: 'input' });
        box.inputs = [socket];
        expect(mock(socket, { box }).index).toBe(0);
    });

    test('input with execInput', () => {
        const socket = Socket.create({ socketType: 'input' });
        box.inputs = [socket];
        box.execInputs = [null];
        expect(mock(socket, { box }).index).toBe(1);
    });

    test('input with output', () => {
        const socket = Socket.create({ socketType: 'input' });
        box.inputs = [socket];
        box.outputs = [null];
        expect(mock(socket, { box }).index).toBe(0);
    });
});

describe('arrows', () => {
    let box: { store: { arrows: ObservableMap } };
    beforeEach(() => {
        box = {
            store: { arrows: new ObservableMap },
        };
    });

    test('by input', () => {
        const socket = mock(Socket.create({ socketType: 'input' }), { box });
        expect(socket.arrows).toHaveLength(0);
        box.store.arrows.set('id', { input: socket });
        expect(socket.arrows).toHaveLength(1);
    });

    test('by output', () => {
        const socket = mock(Socket.create({ socketType: 'output' }), { box });
        expect(socket.arrows).toHaveLength(0);
        box.store.arrows.set('id', { input: socket });
        expect(socket.arrows).toHaveLength(1);
    });
});

describe('x/y', () => {
    const testData = [
        { box: { x: 0, y: 0, width: 0 }, index: 0 },
        { box: { x: -100, y: 100, width: 50 }, index: 0 },
        { box: { x: -100, y: 100, width: 50 }, index: 2 },
        { box: { x: 0, y: 100, width: 20 }, index: 3 },
        { box: { x: -500, y: 0, width: 100 }, index: 5 },
    ];

    const getTest = (socketType: SocketTypeEnum) => () => {
        for (const data of testData) {
            const socket = Socket.create({ socketType });
            const { x, y } = mock(socket, data);
            expect({ x, y }).toMatchSnapshot();
        }
    };

    for (const type of socketTypes) {
        test(type, getTest(type));
    }
});

test('setName', () => {
    const socket = Socket.create({ socketType: 'input' });
    socket.setName(null);
    expect(socket.name).toBe('');
    socket.setName('name');
    expect(socket.name).toBe('name');
});

test('areSocketsCompatible', () => {
    const tests = product(socketTypes, socketTypes, [0, 1], [0, 1]);
    for (const t of tests) {
        const [s1, s2, l1, l2] = t;
        expect(areSocketsCompatible(
            mock(Socket.create({ socketType: s1 }), {arrows: {length: l1}}),
            mock(Socket.create({ socketType: s2 }), {arrows: {length: l2}}),
        )).toMatchSnapshot(t.join(','));
    }
});