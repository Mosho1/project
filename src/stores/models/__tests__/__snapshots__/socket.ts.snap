// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`areSocketsCompatible: execInput,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execInput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execInput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,0,0 1`] = `true`;

exports[`areSocketsCompatible: execOutput,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: execOutput,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: input,output,0,0 1`] = `true`;

exports[`areSocketsCompatible: input,output,0,1 1`] = `true`;

exports[`areSocketsCompatible: input,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: input,output,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execInput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,execOutput,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,0,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,input,1,0 1`] = `true`;

exports[`areSocketsCompatible: output,input,1,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,0,1 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,0 1`] = `false`;

exports[`areSocketsCompatible: output,output,1,1 1`] = `false`;

exports[`x/y execInput 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y execInput 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y execInput 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y execInput 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y execInput 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y execOutput 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y execOutput 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y execOutput 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y execOutput 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y execOutput 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;

exports[`x/y input 1`] = `
Object {
  "x": 18,
  "y": 50,
}
`;

exports[`x/y input 2`] = `
Object {
  "x": -82,
  "y": 150,
}
`;

exports[`x/y input 3`] = `
Object {
  "x": -82,
  "y": 210,
}
`;

exports[`x/y input 4`] = `
Object {
  "x": 18,
  "y": 240,
}
`;

exports[`x/y input 5`] = `
Object {
  "x": -482,
  "y": 200,
}
`;

exports[`x/y output 1`] = `
Object {
  "x": -18,
  "y": 50,
}
`;

exports[`x/y output 2`] = `
Object {
  "x": -68,
  "y": 150,
}
`;

exports[`x/y output 3`] = `
Object {
  "x": -68,
  "y": 210,
}
`;

exports[`x/y output 4`] = `
Object {
  "x": 2,
  "y": 240,
}
`;

exports[`x/y output 5`] = `
Object {
  "x": -418,
  "y": 200,
}
`;
